src/lib/comms/sol-http-common.c

```diff
@@ -74,52 +130,645 @@ sol_http_param_free(struct sol_http_param *params)
         return;
     }
     sol_vector_clear(&params->params);
+    if (params->arena) {
+        sol_arena_del(params->arena);
+        params->arena = NULL;
+    }
+}
+
+SOL_API int
+sol_http_encode_slice(struct sol_buffer *buf, const struct sol_str_slice value)
+{
+    int r;
+    size_t i, last_append;
+    bool escaped;
+
+    SOL_NULL_CHECK(buf, -EINVAL);
+
+    escaped = false;
+    last_append = 0;
+    for (i = 0; i < value.len; i++) {
+        unsigned char c = value.data[i];
+        if (!isalnum(c) && c != '_' && c != '~' && c != '.' && c != '-') {
+            if (!escaped) {
+                sol_buffer_init(buf);
+                escaped = true;
+            }
+            r = sol_buffer_append_slice(buf,
+                SOL_STR_SLICE_STR(value.data + last_append, i - last_append));
+            SOL_INT_CHECK_GOTO(r, < 0, err_exit);
+            last_append = i + 1;
+            r = sol_buffer_append_printf(buf, "%%%02X", c);
+            SOL_INT_CHECK_GOTO(r, < 0, err_exit);
+        }
+    }
+
+    if (!escaped) {
+        sol_buffer_init_flags(buf, (char *)value.data, value.len,
+            SOL_BUFFER_FLAGS_MEMORY_NOT_OWNED);
+        buf->used = buf->capacity;
+    } else if (last_append != value.len) {
+        r = sol_buffer_append_slice(buf,
+            SOL_STR_SLICE_STR(value.data + last_append,
+            value.len - last_append));
+        SOL_INT_CHECK_GOTO(r, < 0, err_exit);
+    }
+
+    return 0;
+
+err_exit:
+    sol_buffer_fini(buf);
+    return r;
+}
+
+SOL_API int
+sol_http_decode_slice(struct sol_buffer *buf,
+    const struct sol_str_slice value)
+{
+    bool decoded;
+    size_t i, last_append;
+    int r;
+
+    SOL_NULL_CHECK(buf, -EINVAL);
+
+    decoded = false;
+    last_append = 0;
+
+    for (i = 0; i < value.len; i++) {
+        unsigned char c = value.data[i];
+        if (c == '%' && value.len - i >= 2 && isxdigit(value.data[i + 1])
+            && value.data[i + 2]) {
```

Comment:
Yep, I forgot the isxdigit() there, I already added locally. And yes it should be > 2